File: tlg/src/main/java/ch/bailu/tlg/InternalContext.java
package ch.bailu.tlg;

import java.io.BufferedOutputStream;
import java.io.IOException;

public class InternalContext {
    protected static final String STATE_FILE = "state";
    protected final static int STATE_FILE_VERSION = 4;

    private final static int MARGIN = 2;
    private final static int TEXT_HEIGHT = 20;


    protected State state;

    protected MatrixWithShape previewMatrix;
    protected MatrixLineManipulator mainMatrix;
    protected TextBox textBox;
    protected Score currentScore;


    private TlgRectangle previewGeometry = new TlgRectangle();
    private TlgRectangle mainGeometry = new TlgRectangle();


    public InternalContext(PlatformContext c) {
        state = new StateInit(this).init(c);
    }


    public synchronized void moveLeft(PlatformContext c) {
        state = state.moveLeft(c);
    }


    public synchronized void moveRight(PlatformContext c) {
        state = state.moveRight(c);
    }


    public synchronized void moveDown(PlatformContext c) {
        state = state.moveDown(c);
    }


    public synchronized void moveTurn(PlatformContext c) {
        state = state.moveTurn(c);
    }


    public synchronized void toggleGrid() {
        mainMatrix.toggleGrid();
    }


    public synchronized void togglePause(PlatformContext c) {
        state = state.togglePause(c);
    }


    public synchronized void startNewGame(PlatformContext c) {
        state = state.startNewGame(c);
    }


    public void writeState(PlatformContext gc) throws IOException {
        BufferedOutputStream output = gc.getOutputStream(STATE_FILE);

        ByteInteger.wrap(STATE_FILE_VERSION).writeState(output);

        previewMatrix.writeState(output);
        mainMatrix.writeState(output);
        currentScore.writeState(output);

        ByteInteger.wrap(getID()).writeState(output);

        output.close();
    }


    protected void setStatusText(String state) {
        String text = "Score: " +
                currentScore.getScore() +
                " Level: " +
                currentScore.getLevel() +
                " | " +
                state;
        textBox.setText(text);
    }


    public synchronized void mainLayout(TlgRectangle r) {
        mainGeometry = r;
        mainMatrix.setPixelGeometry(mainGeometry);
    }


    public synchronized void previewLayout(TlgRectangle r) {
        previewGeometry = r;
        previewMatrix.setPixelGeometry(previewGeometry);
    }


    public synchronized void updatePreview(PlatformContext gc) {
        previewMatrix.update(gc);
    }

    public synchronized void updateAllPreview(PlatformContext gc) {
        updateBackgroundPreview(gc);
        previewMatrix.updateAll(gc);
    }


    public synchronized void updateMain(PlatformContext gc) {
        mainMatrix.update(gc);
    }

    public synchronized void updateAllMain(PlatformContext gc) {
        updateBackgroundMain(gc);
        mainMatrix.updateAll(gc);
    }

    private void updateBackgroundMain(PlatformContext gc) {
        gc.setDirtyRect(mainGeometry);
        gc.drawFilledRectangle(gc.colorBackground(), mainGeometry);
    }


    private void updateBackgroundPreview(PlatformContext gc) {
        gc.setDirtyRect(previewGeometry);
        gc.drawFilledRectangle(gc.colorBackground(), previewGeometry);
    }

    public int getTimerInterval() {
        return state.getTimerInterval();
    }


    public void setHighscoreName(PlatformContext c, String name) {
        state = state.setHighscoreName(c, name);
    }


    public int getID() {
        return state.getID();
    }
}

File: tlg/src/main/kotlin/context/InternalContext.kt
package context

import TlgConfiguration.STATE_FILE
import TlgConfiguration.STATE_FILE_VERSION
import geometry.TlgRectangle
import matrix.MatrixLineManipulator
import matrix.MatrixWithShape
import score.Score
import state.State
import state.StateInit
import java.io.IOException

class InternalContext(c: PlatformContext) {
    var state: State = StateInit(this).init(c)
    var currentScore: Score = Score()
    var previewMatrix: MatrixWithShape = MatrixWithShape(3,3)
    var mainMatrix: MatrixLineManipulator = MatrixLineManipulator(3,3)
    private var previewGeometry: TlgRectangle = TlgRectangle()
    private var mainGeometry: TlgRectangle = TlgRectangle()

    @Synchronized
    fun moveLeft(c: PlatformContext) {
        state = state.moveLeft(c)
    }

    @Synchronized
    fun moveRight(c: PlatformContext) {
        state = state.moveRight(c)
    }

    @Synchronized
    fun moveDown(c: PlatformContext) {
        state = state.moveDown(c)
    }

    @Synchronized
    fun moveTurn(c: PlatformContext) {
        state = state.moveTurn(c)
    }

    @Synchronized
    fun toggleGrid() {
        mainMatrix.toggleGrid()
    }

    @Synchronized
    fun togglePause(c: PlatformContext) {
        state = state.togglePause(c)
    }

    @Synchronized
    fun startNewGame(c: PlatformContext) {
        state = state.startNewGame(c)
    }

    @Throws(IOException::class)
    fun writeState(gc: PlatformContext) {
        val output = gc.getOutputStream(STATE_FILE)
        ByteInteger.wrap(STATE_FILE_VERSION).writeState(output)
        previewMatrix.writeState(output)
        mainMatrix.writeState(output)
        currentScore.writeState(output)
        ByteInteger.wrap(id).writeState(output)
        output.close()
    }

    @Synchronized
    fun mainLayout(r: TlgRectangle) {
        mainGeometry = r
        mainMatrix.pixelGeometry = mainGeometry
    }

    @Synchronized
    fun previewLayout(r: TlgRectangle) {
        previewGeometry = r
        previewMatrix.pixelGeometry = previewGeometry
    }

    @Synchronized
    fun updatePreview(pContext: PlatformContext) {
        previewMatrix.update(pContext)
    }

    @Synchronized
    fun updateAllPreview(pContext: PlatformContext) {
        updateBackgroundPreview(pContext)
        previewMatrix.updateAll(pContext)
    }

    @Synchronized
    fun updateMain(pContext: PlatformContext) {
        mainMatrix.update(pContext)
    }

    @Synchronized
    fun updateAllMain(pContext: PlatformContext) {
        updateBackgroundMain(pContext)
        mainMatrix.updateAll(pContext)
    }

    private fun updateBackgroundMain(gc: PlatformContext) {
        gc.setDirtyRect(mainGeometry)
        gc.drawFilledRectangle(gc.colorBackground(), mainGeometry)
    }

    private fun updateBackgroundPreview(gc: PlatformContext) {
        gc.setDirtyRect(previewGeometry)
        gc.drawFilledRectangle(gc.colorBackground(), previewGeometry)
    }

    val timerInterval: Int
        get() = state.timerInterval

    fun setHighscoreName(pContext: PlatformContext, name: String) {
        state = state.setHighscoreName(pContext, name)
    }

    val id: Int
        get() = state.id
    val stateText: String
        get() = state.toString()
    val level: Int
        get() = currentScore.level
    val score: Int
        get() = currentScore.score

}
